code/modules/mob/living/silicon/robot/robot_modules.dm

```diff
@@ -6,77 +6,141 @@
-	QDEL_NULL(emag)
+	QDEL_LIST(basic_modules)
+	QDEL_LIST(default_modules)
+	QDEL_LIST(emag_modules)
+	QDEL_LIST(storages)
+	QDEL_LIST(special_rechargables)
 	return ..()
 
-// By default, all robots will get the items in this proc, unless you override it for your specific module. See: ../robot_module/drone
-/obj/item/robot_module/proc/add_default_robot_items()
-	modules += new /obj/item/flash/cyborg(src)
-
+// Here for admin debugging purposes only.
 /obj/item/robot_module/proc/fix_modules()
 	for(var/obj/item/I in modules)
 		I.flags |= NODROP
 		I.mouse_opacity = MOUSE_OPACITY_OPAQUE
-	if(emag)
-		emag.flags |= NODROP
-		emag.mouse_opacity = MOUSE_OPACITY_OPAQUE
-
-/obj/item/robot_module/proc/respawn_consumable(mob/living/silicon/robot/R)
-	if(!stacktypes || !stacktypes.len)
-		return
-
-	var/stack_respawned = 0
-	for(var/T in stacktypes)
-		var/O = locate(T) in modules
-		var/obj/item/stack/S = O
-
-		if(!S)
-			S = new T(src)
-			modules += S
-			S.amount = 1
-			stack_respawned = 1
-
-		if(S && S.amount < stacktypes[T])
-			S.amount++
-	if(stack_respawned && istype(R) && R.hud_used)
-		R.hud_used.update_robot_modules_display()
 
-/obj/item/robot_module/proc/rebuild()//Rebuilds the list so it's possible to add/remove items from the module
-	var/list/temp_list = modules
+/// Returns a `robot_energy_strage` datum of type `storage_type`. If one already exists, it returns that one, otherwise it create a new one.
+/obj/item/robot_module/proc/get_or_create_estorage(storage_type)
+	for(var/datum/robot_energy_storage/S in storages)
+		if(istype(S, storage_type))
+			return S
+	return new storage_type(src)
+
+/// Adds the item `I` to our `modules` list, gives it NODROP, and sets up any energy storage if its a stack.
+/obj/item/robot_module/proc/add_module(obj/item/I, requires_rebuild)
+	if(istype(I, /obj/item/stack))
+		var/obj/item/stack/S = I
+
+		if(S.energy_type)
+			S.source = get_or_create_estorage(S.energy_type)
+
+		if(S && S.source)
+			S.is_cyborg = TRUE
+
+	if(I.loc != src)
+		I.forceMove(src)
+
+	modules += I
+	I.flags |= NODROP
+	I.mouse_opacity = MOUSE_OPACITY_OPAQUE
+
+	if(requires_rebuild)
+		rebuild_modules()
+	return I
+
+/// Builds the usable module list from the modules we have in `default_modules`, `basic_modules` and `emag_modules`
+/obj/item/robot_module/proc/rebuild_modules()
+	var/mob/living/silicon/robot/R = loc
+	R.uneq_all()
 	modules = list()
-	for(var/obj/O in temp_list)
-		if(O)
-			modules += O
 
+	for(var/obj/item/I in default_modules)
+		add_module(I, FALSE)
+
+	for(var/obj/item/I in basic_modules)
+		add_module(I, FALSE)
+
+	if(R.emagged)
+		for(var/obj/item/I in emag_modules)
+			add_module(I, FALSE)
+
+	if(R.hud_used)
+		R.hud_used.update_robot_modules_display()
+
+/// Handles the recharging of all borg stack items and any items which are in the `special_rechargables` list.
+/obj/item/robot_module/proc/recharge_consumables(mob/living/silicon/robot/R, coeff = 1)
+	for(var/datum/robot_energy_storage/st in storages)
```

Comment:
Same here and the one below.
Can this list have anything other than the filtered var type?

Reply:
Nope it should always contain those datums.